alpa/collective/collective_group/xla_nccl_collective_group.py

```diff
@@ -0,0 +1,393 @@
+                                   comm_key,
+                                   my_gpu_idx,
+                                   peer_rank,
+                                   peer_gpu_idx,
+                                   nccl_uid=None):
+        """Create or retrieve an NCCL communicator for p2p tasks.
+
+        Args:
+            comm_key (str): communicator key.
+            my_gpu_idx (int): the gpu index on the current process.
+            peer_rank (int): the rank of the destination process.
+            peer_gpu_idx (int): the gpu index on the peer process.
+        Returns:
+            communicator
+        """
+        # pylint: disable=unused-argument
+        if not comm_key:
+            raise RuntimeError("Got empty communicator key.")
+
+        # TODO(Hao): lock the _dev_comm_map here.
+        if comm_key in self._dev_comm_map:
+            return self._dev_comm_map[comm_key]
+
+        # Note (Hao): This is a bit complex so I decide to take a note here.
+        # Here we need to consider three cases:
+        # Case 1: src_rank != dst_rank, hence the send and recv happen on
+        # different process (actors/tasks); each process makes independent
+        # collective calls and manages corresponding communicators.
+        # Case 2: src_rank == dst_rank, src_gpu_idx == dst_gpu_idx; for
+        # this case, we simply throw a RuntimeError;
+        # Case 3: src_rank == dst_rank, src_gpu_idx != dst_gpu_idx, which
+        # means the send and recv will be called on the same process. We
+        # DO NOT support this case for now. We need to properly scope:
+        # (1) communicators creation, and
+        # (2) send/recv calls
+        # using groupStart(（ and groupEnd() calls to avoid deadlocks.
+        if self.rank < peer_rank:
+            my_p2p_rank = 0
+        elif self.rank > peer_rank:
+            my_p2p_rank = 1
+        else:
+            raise RuntimeError(
+                "Send and recv happens on the same process! "
+                "alpa.collective does not support this case as of now. "
+                "Alternatively, consider doing GPU to GPU memcpy?")
+        group_key = self._generate_group_key(comm_key)
+        if my_p2p_rank == 0:
+            if nccl_uid is None:
+                nccl_uid = self._generate_nccl_uid(group_key)
+        else:
+            if nccl_uid is None:
+                rendezvous = Rendezvous(group_key)
+                rendezvous.meet(timeout_s=3000)
+                nccl_uid = rendezvous.get_nccl_id()
+                # Recycle the NCCLUniqueIDStore named actor *pro-activately* to
+                # avoid named actor leak.
+                if rendezvous.get_access_counter() == 2:
+                    logger.debug(
+                        "NCCLUniqueID has been broadcasted. The "
+                        "NCCLUniqueIDStore will go out of context and be "
+                        "destroyed.")
+                    rendezvous.destroy_store()
```

Comment:
Again, do not copy code. Especially when the copied logic seems complicated.

Reply:
Thx! I will delete all cupy nccl in another PR when nccl XLA works for everyone.